packages/devtools_app/lib/src/memory/flutter/memory_chart.dart

```diff
@@ -73,32 +71,28 @@ class MemoryChartState extends State<MemoryChart> with AutoDisposeMixin {
     if (newController == controller) return;
     controller = newController;
 
+    controller.memoryTimeline.image = _img;
+
+    _setupChart();
+
     cancel();
 
-    controller.addMemorySourceListener(() {
+    // Update the chart when the memorySource changes.
+    addAutoDisposeListener(controller.memorySourceNotifier, () {
       setState(() {
-        // The memory source has changed, clear all plotted values.
-        _used.clear();
-        _capacity.clear();
-        _externalHeap.clear();
-
-        // Reset the chart
-        _setupChart();
-
-        if (controller.offline)
-          // Replot the entire offline data.
-          processMemoryLogFileData();
-        else
-          // Continue plot the live heap samples being
-          // received.
-          processLiveData(true);
+        controller.updatedMemorySource();
+
+        // Plot all offline or online data (based on memorySource).  If
+        // not offline then all new heap samples will be plotted as it
+        // appears via the sampleAddedNotifier.
+        _processAndUpdate(true);
       });
     });
 
-    // Process and plot each heap sample as it is received.
+    // Plot each heap sample as it is received.
     addAutoDisposeListener(
       memoryTimeline.sampleAddedNotifier,
-      processLiveData,
+      _processAndUpdate,
```

Comment:
same here - instead of updating from the UI code, just re-process the data in the controller when a sample is added